### PR TITLE
Add changes to support all the grant types

### PR DESCRIPTION
**Pull request by kalanika**
[1] - https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop-02#section-5

### CODE REVIEW

File: component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java [hunk 1]
```diff
@@ -241,4 +243,18 @@ private boolean validateDPoPHeader(Object request, TokenBinding tokenBinding) th
...
+        if (supportedGrantTypes == null) {
+            synchronized (DPoPBasedTokenBinder.class) {
+                if (supportedGrantTypes == null) {
+                    Set<String> allowedGrantSet =
+                            OAuthServerConfiguration.getInstance().getSupportedGrantTypes().keySet();
+                    supportedGrantTypes = new ArrayList<>(allowedGrantSet);
+                }
+            }
+        }
+        return supportedGrantTypes.toArray(new String[supportedGrantTypes.size()]);
```
Comment (chamathns) [inline component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java:258]: Can't we convert the set to an array straight away?

Comment (kalanika) [reply component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java]: Fixed

File: component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java [hunk 2]
```diff
@@ -76,7 +78,7 @@ public String getBindingType() {
     @Override
     public List<String> getSupportedGrantTypes() {
 
-        return Collections.unmodifiableList(supportedGrantTypes);
+        return new ArrayList<>(Arrays.asList(getAllGrantTypes()));
```
Comment (piraveena) [inline component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java:81]: How can we set all the grant types as supported? We are supporting different grant types OOTB. Did we test DPoP with all the grant types

Comment (kalanika) [reply component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java:81]: I have tested DPoP with the below grant types.

1. Authorization Code Grant.
2. Password Grant.
3. Client Credentials Grant.
4. Refresh Token Grant.
5. Custom grant type[1].

[1] - https://is.docs.wso2.com/en/latest/learn/writing-a-custom-oauth-2.0-grant-type/#customizing-an-existing-grant-type

**Review by chamathns — APPROVED** [review]

File: component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java [hunk 3]
```diff
@@ -50,11 +52,11 @@
 
     private static final String BINDING_TYPE = "DPoP";
     private static final Log log = LogFactory.getLog(DPoPBasedTokenBinder.class);
-    private final List<String> supportedGrantTypes = Arrays.asList(GrantTypes.AUTHORIZATION_CODE, GrantTypes.PASSWORD,
-            GrantTypes.CLIENT_CREDENTIALS, GrantTypes.REFRESH_TOKEN);
+    static Set<String> supportedGrantTypesSet = null;
```
Comment (chamathns) [inline component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java:55]: shall we use `Collections.<String>emptySet();`here?

Comment (kalanika) [reply component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/token/binder/DPoPBasedTokenBinder.java]: fixed

**Comment by kalanika** [comment]
Resolves wso2/product-is#9102

**Pull request merged**